[DA-3042] Remote ID Verification

ayazamlani opened the pull request:
## Resolves *[ticket DA-3042]* 
https://precisionmedicineinitiative.atlassian.net/browse/DA-3042

## Description of changes/additions

** Adding Remote ID Verification logic to accept from questionnaire responses and update PS table accordingly

## Tests
- [x] unit tests


[DA-3042]: https://precisionmedicineinitiative.atlassian.net/browse/DA-3042?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ayazamlani committed:
added remote id verification classes and the payloads for unit tests

ayazamlani committed:
Test remote identity verified. Fixed an Enum.
added user info config to get participantOrigin set to vibrent. Changed remoteIdVerifiedOn from DateTime to Date. Added REMOTE_ID_VERIFIED logic to the questionnaire_response_dao.py Now encountering KeyError: 'remoteIdVerifiedOn'

ayazamlani committed:
Fixed logic for questionnaire response dao and wrote a sad path test

ayazamlani committed:
cleanup

ayazamlani committed:
reset the config after each test, asserts easier to read, added remoteIdVerificationStatus to participant_summary_default_values. Changed in participant_summary.py remoteIdVerificationStatus to nullable

ayazamlani committed:
Merge branch 'devel' into aa/DA-3042

ayazamlani committed:
Merge branch 'devel' into aa/da-3042

ayazamlani committed:
adding alembic migration file

ayazamlani committed:
adding checks to remote IDV tests to look at the questionnaire response table and make sure data is accurate

ayazamlani committed:
codacy checks

ayazamlani committed:
additional codacy checks

ayazamlani committed:
renaming tests

brianmkimmel commented on rdr_service/dao/questionnaire_response_dao.py:
```diff
@@ -931,6 +939,18 @@ def _update_participant_summary(
 
         dna_program_consent_update_code = config.getSettingJson(config.DNA_PROGRAM_CONSENT_UPDATE_CODE, None)
 
+        if 'verified' in remote_id_info.keys() and 'verified_on' in remote_id_info.keys():
+            if remote_id_info['verified'] == 1:
+                # is the participant origin always going to be the same? if not, where can I pull this from
+                participant_summary.remoteIdVerificationOrigin = participant_summary.participantOrigin
+                participant_summary.remoteIdVerificationStatus = RemoteIdVerificationStatus.TRUE
+                participant_summary.remoteIdVerifiedOn = remote_id_info['verified_on']
+        elif 'verified' in remote_id_info.keys():
```
You don't need to use .keys() here, the conditions in this if elif block can be simplified: 
```suggestion
        if 'verified' in remote_id_info and 'verified_on' in remote_id_info:
            if remote_id_info['verified'] == 1:
                # is the participant origin always going to be the same? if not, where can I pull this from
                participant_summary.remoteIdVerificationOrigin = participant_summary.participantOrigin
                participant_summary.remoteIdVerificationStatus = RemoteIdVerificationStatus.TRUE
                participant_summary.remoteIdVerifiedOn = remote_id_info['verified_on']
        elif 'verified' in remote_id_info:
```

ayazamlani committed:
Update rdr_service/dao/questionnaire_response_dao.py

simplifying conditions in the if else statement

Co-authored-by: Brian Kimmel <brian.kimmel@vumc.org>

dtharpeuno commented on rdr_service/model/participant_summary.py:
```diff
@@ -1555,8 +1557,23 @@ def enrollmentSiteId(cls):
+    )
+    """ A true or false value that indicates whether the ID Verification is Remote  """
+
+    remoteIdVerifiedOn = Column("remote_id_verified_on", Date)
```
Think this should be `UTCDateTime`, you will need to edit the migration as well.

dtharpeuno replied:
Maybe we can simplify this and just use:
`remoteIdVerificationStatus = Column("remote_id_verification_status", Boolean, default=0)`
 instead of leveraging another enum

dtharpeuno replied:
Still needs to be changed to `UTCDateTime`

dtharpeuno commented on rdr_service/participant_enums.py:
```diff
@@ -1021,3 +1021,8 @@ class WorkbenchResearcherYesNoPreferNot(messages.Enum):
     PREFER_NOT_TO_ANSWER = 3
+
+
+class RemoteIdVerificationStatus(messages.Enum):
+    FALSE = 0
+    TRUE = 1
```
Dont think we need this enum class for booleans, we can just use 0/1 in the model attribute and the API will render it correctly.

dtharpeuno commented on tests/api_tests/test_questionnaire_response_api.py:
```diff
@@ -1707,6 +1707,73 @@ def test_questionnaire_life_functioning_survey(self):
+        self.assertEqual(response_obj.answers[0].valueDecimal, 0)
+        self.assertEqual(response_obj.answers[1].valueDate, datetime.date(2022, 11, 30))
+        # Get request from API to assert information is accurate
+        summary = self.send_get("Participant/%s/Summary" % participant_id)
+        self.assertNotIn('remoteIdVerificationOrigin', summary.keys())
+        self.assertNotIn('remoteIdVerifiedOn', summary.keys())
+        self.assertEqual(summary['remoteIdVerificationStatus'], 'UNSET')
+        # Reset Config back to original user client ID
+        user_info['example@example.com']['clientId'] = original_user_client_id
+        config.override_setting(config.USER_INFO, user_info)
+
```
Think we discussed moving this to `test_participant_summary_api.py`.
And leaning into the dao to test the data that is getting set.

ayazamlani replied:
Yes, Sry, I am trying to figure out how to unpack and submit the response answers in the participant summary test file using the code answers, date answers, etc. I left these asserts in here for now.

dtharpeuno commented on tests/api_tests/test_questionnaire_response_api.py:
```diff
@@ -1707,6 +1707,73 @@ def test_questionnaire_life_functioning_survey(self):
+        self.assertEqual(response_obj.answers[0].valueDecimal, 1)
+        self.assertEqual(response_obj.answers[1].valueDate, datetime.date(2022, 11, 30))
+        # Get request from API to assert information is accurate
+        summary = self.send_get("Participant/%s/Summary" % participant_id)
+        self.assertEqual(summary['remoteIdVerificationOrigin'], 'vibrent')
+        self.assertEqual(summary['remoteIdVerificationStatus'], 'TRUE')
+        self.assertEqual(summary['remoteIdVerifiedOn'], '2022-11-30')
+        # Reset Config back to original user client ID
+        user_info['example@example.com']['clientId'] = original_user_client_id
+        config.override_setting(config.USER_INFO, user_info)
```
Same here in moving this to  `test_participant_summary_api.py`. 
And leaning into the dao to test the data that is getting set.

dtharpeuno commented on rdr_service/dao/questionnaire_response_dao.py:
```diff
@@ -931,6 +939,18 @@ def _update_participant_summary(
 
         dna_program_consent_update_code = config.getSettingJson(config.DNA_PROGRAM_CONSENT_UPDATE_CODE, None)
 
+        if 'verified' in remote_id_info and 'verified_on' in remote_id_info:
```
:nit 
`if 'verified' and 'verified_on' in remote_id_info:`

ayazamlani committed:
updated ps columns and alembic file to match
removed RemoteIdStatus Enum & used a Boolean
changed RemoteIdVerifiedOn from Date to UTCDateTime
refactored load_remote_measurement_json to load_questionnaire_response_json
updated qr_tests to lean on the dao
moved ps asserts from test_questionnaire_response_api.py to test_participant_summary_api.py

ayazamlani committed:
Merge remote-tracking branch 'origin/aa/DA-3042' into aa/DA-3042

# Conflicts:
#	rdr_service/dao/questionnaire_response_dao.py

dtharpeuno commented on tests/api_tests/test_questionnaire_response_api.py:
```diff
@@ -1707,6 +1707,70 @@ def test_questionnaire_life_functioning_survey(self):
+    def test_remote_identity_verified(self):
+        """ Test to see if a remote ID verification True Response saves successfully """
+        # Set up user config to have client set to vibrent
+        user_info = config.getSettingJson(config.USER_INFO)
+        original_user_client_id = user_info['example@example.com']['clientId']
+        user_info['example@example.com']['clientId'] = 'vibrent'
+        config.override_setting(config.USER_INFO, user_info)
```
Shouldn't need this for the tests, you can do the POST with the default account.

ayazamlani replied:
I see, thanks for catching this!

dtharpeuno commented on tests/api_tests/test_questionnaire_response_api.py:
```diff
@@ -1707,6 +1707,70 @@ def test_questionnaire_life_functioning_survey(self):
+        qr = self.dao.get_with_children(response['id'])
+        self.assertEqual(qr.answers[0].valueDecimal, 1)
+        self.assertEqual(qr.answers[1].valueDate, datetime.date(2022, 11, 30))
+        # Reset Config back to original user client ID
+        user_info['example@example.com']['clientId'] = original_user_client_id
+        config.override_setting(config.USER_INFO, user_info)
```
Can remove this also

dtharpeuno commented on tests/api_tests/test_questionnaire_response_api.py:
```diff
@@ -1707,6 +1707,70 @@ def test_questionnaire_life_functioning_survey(self):
+    def test_remote_identity_not_verified(self):
+        """ Test to see if a remote ID verification False Response saves successfully """
+        # Set up user config to have client set to vibrent
+        user_info = config.getSettingJson(config.USER_INFO)
+        original_user_client_id = user_info['example@example.com']['clientId']
+        user_info['example@example.com']['clientId'] = 'vibrent'
+        config.override_setting(config.USER_INFO, user_info)
```
Shouldn't need this for the tests, you can do the POST with the default account.

dtharpeuno commented on tests/api_tests/test_questionnaire_response_api.py:
```diff
@@ -1707,6 +1707,70 @@ def test_questionnaire_life_functioning_survey(self):
+        qr = self.dao.get_with_children(response['id'])
+        self.assertEqual(qr.answers[0].valueDecimal, 0)
+        self.assertEqual(qr.answers[1].valueDate, datetime.date(2022, 11, 30))
+        # Reset Config back to original user client ID
+        user_info['example@example.com']['clientId'] = original_user_client_id
+        config.override_setting(config.USER_INFO, user_info)
```
Can remove this also

dtharpeuno commented on tests/api_tests/test_participant_summary_api.py:
```diff
@@ -4424,3 +4426,59 @@ def test_updated_since_parameter(self):
+    def test_remote_identity_not_verified(self):
+        """ Test to see if a remote ID verification False Response saves successfully """
+        # Set up user config to have client set to vibrent
+        user_info = config.getSettingJson(config.USER_INFO)
+        original_user_client_id = user_info['example@example.com']['clientId']
+        user_info['example@example.com']['clientId'] = 'vibrent'
+        config.override_setting(config.USER_INFO, user_info)
```
Shouldn't need this for the tests, you can do the POST with the default account.

dtharpeuno commented on tests/api_tests/test_participant_summary_api.py:
```diff
@@ -4424,3 +4426,59 @@ def test_updated_since_parameter(self):
+    def test_remote_identity_verified(self):
+        """ Test to see if a remote ID verification True Response saves successfully """
+        # Set up user config to have client set to vibrent
+        user_info = config.getSettingJson(config.USER_INFO)
+        original_user_client_id = user_info['example@example.com']['clientId']
+        user_info['example@example.com']['clientId'] = 'vibrent'
+        config.override_setting(config.USER_INFO, user_info)
```
Shouldn't need this for the tests, you can do the POST with the default account.

dtharpeuno commented on tests/api_tests/test_participant_summary_api.py:
```diff
@@ -4424,3 +4426,59 @@ def test_updated_since_parameter(self):
+        self.assertNotIn('remoteIdVerificationOrigin', summary)
+        self.assertNotIn('remoteIdVerifiedOn', summary)
+        self.assertEqual(summary['remoteIdVerificationStatus'], False)
+        # Reset Config back to original user client ID
+        user_info['example@example.com']['clientId'] = original_user_client_id
+        config.override_setting(config.USER_INFO, user_info)
```
Can remove this also

dtharpeuno commented on tests/api_tests/test_participant_summary_api.py:
```diff
@@ -4424,3 +4426,59 @@ def test_updated_since_parameter(self):
+        self.assertEqual(summary['remoteIdVerificationOrigin'], 'vibrent')
+        self.assertEqual(summary['remoteIdVerificationStatus'], True)
+        self.assertEqual(summary['remoteIdVerifiedOn'], '2022-11-30T00:00:00')
+        # Reset Config back to original user client ID
+        user_info['example@example.com']['clientId'] = original_user_client_id
+        config.override_setting(config.USER_INFO, user_info)
```
Can remove this also

dtharpeuno commented on rdr_service/dao/questionnaire_response_dao.py:
```diff
@@ -891,6 +894,16 @@ def _update_participant_summary(
                             participant_summary.consentForElectronicHealthRecordsAuthored = authored
                             participant_summary.consentForElectronicHealthRecordsTime = questionnaire_response.created
+                    elif code.value.lower() == REMOTE_ID_VERIFIED_ON_CODE:
+                        utc_datetime = datetime(
+                            answer.valueDate.year,
+                            answer.valueDate.month,
+                            answer.valueDate.day,
+                            tzinfo=pytz.timezone('UTC')
+                                            )
```
Do we need to do this? I think the value will be stored as UTC time on insertion. I haven't seen anywhere else where we do this.

kenny-skaggs replied:
Right, I don't think we need to do this either. The UTCDateTime datatype (https://github.com/all-of-us/raw-data-repository/blob/devel/rdr_service/model/utils.py#L83) will convert it to UTC timezone. For future knowledge, you'll want to be careful when making that conversion. It looks like the code here wouldn't necessarily convert, but instead replaces the timezone information provided.

ayazamlani replied:
I am not seeing anywhere else we do this, however when I leave just the answer.valueDate I am receiving this error:
`sqlalchemy.exc.StatementError: (builtins.AttributeError) 'datetime.date' object has no attribute 'tzinfo'`

kenny-skaggs replied:
Ah, ok. Sorry I hadn't really thought through that we're storing just the date here. I think the verified_on column should just be a Date datatype then (rather than a variation of Datetime). That way we can store just the date value sent and not have to do anything with timezones.

kenny-skaggs commented on rdr_service/dao/questionnaire_response_dao.py:
```diff
@@ -931,6 +944,18 @@ def _update_participant_summary(
 
         dna_program_consent_update_code = config.getSettingJson(config.DNA_PROGRAM_CONSENT_UPDATE_CODE, None)
 
+        if 'verified' and 'verified_on' in remote_id_info:
```
I know there was a comment to change it to this, but the logic here doesn't look like what was meant. This `if` condition is evaluating to True if `'verified_on' in remote_id_info` is true. This is not checking if 'verified' and 'verified_on' are in `remote_id_info`. Instead it's testing if the 'verified' string is truthy (which will always be True since non-empty strings are truthy).

ayazamlani replied:
Nice catch, Thanks!

ayazamlani committed:
Changed remoted_id_verified_on back to Date from UTCDateTime
refactored function names to meet codacy check
removed user configurations in tests and now using default participant.

ayazamlani committed:
resolving dao_test errors

ayazamlani committed:
update test_remote_id_not_verified assert

ayazamlani commented:
Thank you @dtharpeuno @kenny-skaggs @brianmkimmel for feedback. Resubmitting for approval.

dtharpeuno commented on rdr_service/alembic/versions/7d92a6f2ff9b_remote_id_fieleds.py:
```diff
@@ -0,0 +1,68 @@
+                                                   sa.Boolean(),
+                                                   nullable=True))
+    op.add_column('participant_summary', sa.Column('remote_id_verified_on',
+                                                   sa.Date(),
```
Migration still needs to be updated as well

dtharpeuno commented on tests/dao_tests/test_questionnaire_response_dao.py:
```diff
@@ -1355,7 +1355,7 @@ def test_insert_qr_three_times(self):
             enrollmentStatusParticipantV3_1Time=datetime.datetime(2016, 1, 2),
             healthDataStreamSharingStatusV3_1=DigitalHealthSharingStatusV31.NEVER_SHARED,
-            aian=0
+            aian=False
```
Can we revert this so it matches entries in the rest of the repo.

ayazamlani replied:
My tests fail when it is set to 0, not sure how others are passing? Did I merge this incorrectly or something from a recent PR as I did not mess with aian on this ticket.

ayazamlani replied:
checking on this again now

ayazamlani replied:
Fixed, so sorry about this accident! ty for catching this @dtharpeuno

dtharpeuno commented on rdr_service/dao/questionnaire_response_dao.py:
```diff
@@ -931,6 +938,18 @@ def _update_participant_summary(
 
+        if 'verified' in remote_id_info and 'verified_on' in remote_id_info:
+            if remote_id_info['verified'] == 1:
+                # is the participant origin always going to be the same? if not, where can I pull this from
```
Can we remove this comment?

ayazamlani committed:
cleanup

ayazamlani committed:
reverting AIAN back to 0 in expected_ps3 of test_questionnaire_response_dao.py

ayazamlani committed:
reverting AIAN back to 0 in test_insert_qr_three_times of test_questionnaire_response_dao.py

ayazamlani committed:
codacy check cleanup

dtharpeuno reviewed (approved):
👍
